lib/rucio/daemons/rsedecommissioner/profiles/generic.py

```diff
@@ -0,0 +1,416 @@
+                logger(logging.WARNING, '(%s) %s', rse['rse'], str(error))
+            else:
+                updated = True
+
+            reap_pending = False
+
+        if updated:
+            num_updated += 1
+
+        if reap_pending:
+            num_waiting_reap += 1
+
+    logger(logging.INFO, '(%s) %s replicas have been updated. %s replicas are pending deletion.'
+           ' %s replicas remain unhandled.',
+           rse['rse'], num_updated, num_waiting_reap,
+           num_processed - num_updated - num_waiting_reap)
+
+
+# Condition functions
+
+def _is_being_deleted(
+    rule: dict[str, Any],
+    rse: dict[str, Any],
+    *,
+    logger: Callable[..., None] = logging.log
+) -> bool:
+    """Check if the rule is already set to be deleted."""
+    if rule['expires_at'] is not None and rule['expires_at'] < datetime.utcnow():
+        logger(logging.DEBUG,
+               '(%s) Rule %s for %s:%s is bound for deletion',
+               rse['rse'], rule['id'], rule['scope'], rule['name'])
+        return True
+    return False
+
+
+def _has_trivial_rse_expression(
+    rule: dict[str, Any],
+    rse: dict[str, Any],
+    *,
+    logger: Callable[..., None] = logging.log
+) -> bool:
+    """Check for a trivial rse_expression."""
+    return rule['rse_expression'] == rse['rse']
+
+
+def _has_last_replica_on_rse(
+    rule: dict[str, Any],
+    rse: dict[str, Any],
+    *,
+    logger: Callable[..., None] = logging.log
+) -> bool:
+    """Check if the only replica is on the RSE."""
+    # Check the list of RSEs that the replicas locked by this rule reside on.
+    try:
+        locks = get_replica_locks_for_rule_id(rule['id'])
+    except NoResultFound:
+        # No replica is locked to begin with -> treat as having the last replica on this RSE
+        return True
+
+    replica_rse_ids = set(lock['rse_id'] for lock in locks)
+
+    return len(replica_rse_ids) == 1 and list(replica_rse_ids)[0] == rse['id']
+
+
+def _has_child_rule_id(
+    rule: dict[str, Any],
+    rse: dict[str, Any],
+    *,
+    logger: Callable[..., None] = logging.log
+) -> bool:
+    """Check for non-empty child_rule_id."""
+    if rule['child_rule_id']:
+        logger(logging.DEBUG, '(%s) Rule %s for %s:%s is being moved',
+               rse['rse'], rule['id'], rule['scope'], rule['name'])
+        return True
+    return False
+
+
+# Action functions
+
+def _count_as_processed(
+    rule: dict[str, Any],
+    rse: dict[str, Any],
+    *,
+    logger: Callable[..., None] = logging.log
+) -> bool:
+    """Do nothing but increment the limit checker."""
+    return True
+
+
+def _delete_rule(
+    rule: dict[str, Any],
+    rse: dict[str, Any],
+    *,
+    logger: Callable[..., None] = logging.log
+) -> bool:
+    """Delete the rule."""
+    logger(logging.DEBUG, '(%s) Deleting rule %s for %s:%s',
+           rse['rse'], rule['id'], rule['scope'], rule['name'])
+
+    try:
+        update_rule(rule['id'], {'lifetime': 0})
```

Comment:
Please remind me, did we discuss the case of `rule['locked'] is True`? These won’t be deleted and the Decommissioner shouldn’t touch them (the motivation is that unlocked requires manual check and intervention by an operator), but maybe you should log a warning in `_is_being_deleted()`.

Reply:
I added a new condition function to check for `rule['locked']`. The corresponding action should be to call for attention (i.e. manual intervention), and for that I decided that the current "action result is None or True or False" scheme was not expressive enough. So I instead introduced an Enum called HandlerOutcome, which can take values UNTOUCHED, REMOVED, and NEED_ATTENTION. If the rule is locked, the returned outcome is NEED_ATTENTION.

Reply:
I very much prefer this design, but maybe you need to log something in `_call_for_attention()`? It could be using the `DEBUG` level, but there should be something somewhere to aid the operators to identify the rules that necessitate manual action. Excuse me if you’re already doing this somewhere and I missed it.

Reply:
That's a very good point! I added INFO-level messages in `_is_locked()` and also after all the handlers are exhausted in `DecommissioningProfile.process()`.